doc/glossary.rst

```diff
@@ -0,0 +1,842 @@
+.. currentmodule:: sklearn
+
+.. _glossary:
+
+=========================================
+Glossary of Common Terms and API Elements
+=========================================
+
+This glossary hopes to definitively represent the tacit and explicit
+conventions applied in Scikit-learn and its API, while providing a reference
+for users and contributors. It aims to describe the concepts and either detail
+their corresponding API or link to other relevant parts of the documentation
+which do so.
+
+General Concepts
+================
+
+.. glossary::
+
+    affinity
+        TODO
+
+    API
+        TODO
+
+        The specific interfaces for estimators implemented in Scikit-learn and
+        the generalized conventions across types of estimators as described in
+        this glossary.
+
+        Conventions about what's public and what's not.
+
+    array-like
+
+        The most common data format for *input* to Scikit-learn estimators and
+        functions, array-like is any type object for which
+        :func:`numpy.asarray` will produce an array of appropriate shape
+        (usually 1 or 2-dimensional) of appropriate dtype (usually numeric).
+
+        This includes:
+
+        * a numpy array
+        * a list of numbers
+        * a list of length-k lists of numbers for some fixed length k
+        * a :class:`pandas.DataFrame` with all columns numeric
+        * a numeric :class:`pandas.Series`
+
+        It excludes:
+
+        * a :term:`sparse matrix`
+        * an iterator
+        * a generator
+
+        Note that *output* from scikit-learn estimators and functions (e.g.
+        predictions) should generally be arrays or sparse matrices, or lists
+        thereof. An estimator where ``predict()`` returns a list or a Series is
+        not valid.
+
+    attribute
+    attributes
+
+        We mostly use *attribute* to refer to how model information is stored
+        on an estimator during fitting.  Public attributes on Scikit-learn
+        estimators conventionally begin with an alphabetic character and end
+        with a single underscore (``_``) -- for instance ``coef_`` -- are
+        available as attributes (in the Python object sense) of a
+        :term:`fitted` estimator, and are described in the estimator's
+        docstring.
+
+        The information stored in attributes are usually either: sufficient
+        statistics used for prediction or transformation; or diagnostic data,
+        such as :term:`feature_importances_`.
+
+        Common attributes are listed :ref:`below <glossary_attributes>`.
+
+    backwards compatibility
+        TODO
+
+        We may sometimes assume that all optional parameters other than X and y
+        are passed as keyword arguments only and may be positionally reordered.
+
+        See also :term:`deprecation`.
+
+    categorical feature
+        TODO
+
+    classifier
```

Comment:
Should `classifier` not be moved to the `Class APIs and Estimator Types` section?

It feels weird that it is not in the same section as regressor but maybe there is a good reason for this.

Reply:
No good reason, just a refactoring error.